text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+field with an async destructor having to implement a "forwarding async drop" to
+ensure the field's destructor is called.
+
+That said, there are alternatives along this continuum that make it nicer to
+implement an async destructor while guaranteeing that it will be called in less
+circumstances, and we should document them.
+
+## Have no async drop glue
+
+The first step would be to have no async drop glue: we would only drop values
+when they themselves are dropped in an async context, not when types which have
+a field of that value are dropped in an async context.
+
+This would enable us to make two modifications to the design:
+* `poll_drop_ready` could take self by `Pin`
+* `poll_drop_ready` would not need to have fused semantics
+
+Because there's automatic drop glue generated for fields, we would not need to
+automatically pin project to them to have them take self by pin. And because
+there's no drop glue that needs to be stateless, we would not need fused
+semantics.
+
+However, this would make it the responsibility of the author of every type
+which has a field which implements `poll_drop_ready` to implement the drop glue
+as a part of their own `poll_drop_ready` implementation. And if we eliminate
+fused semantics and make them take self by pin, that drop glue also cannot be
+as simple to implement as the drop glue this RFC proposes we auto generate. We
+would essentially place all of the burden on everyone who wants to guarantee
+the `poll_drop_ready` of their field gets called.
+
+Note that "any type which has a field which implements `poll_drop_ready` means,
+in practice, *any generic type*, because any generic parameter could possibly
+implement `poll_drop_ready`. This would be incredibly burdensome across the
+ecosystem.
+
+## Have no virtual async drop, more weirdness around auto traits
+
+The ideal and most immediately obvious API for an async destructor is something
+like this:
+
+```rust
+trait AsyncDrop {
+    async fn drop(&mut self);
+}
+```
+
+However, this introduces several problems because it would allow the body of
+the async fn to introduce arbitrary additional state not contained in the type
+itself. This would cause the same problems that non-fused semantics would cause
+for drop glue, but it would also cause even more problems.
+
+The first is that we would not be able to async drop a trait object. When a
+trait object is dropped, we dynamically dispatch the call to the destructor of
+the concrete type that backs that object, through the vtable of the trait
+object. However, this definition of async drop would require some place to put
+the state of the future the async call returns, and the layout of each future
+for each concrete type would be different and unknown. It would be functionally
+equivalent to having a trait object with an unknown associated type, which is
+not allowed for exactly this reason. It would not be possible to support async
```

Comment:
That's definitely a drawback. But `async fn` in traits and trait objects is in general a problem that still requires a solution - I'm not sure whether drop is special in that regard.

On the flip side we want people to write asychronous functions and not manual state-machines - because the latter are hard! That's why invented async/await. Inside drop we might want to call functions that are defined as async fn, e.g. for performing the async database rollback, which might lock an async semaphore, and then uses an async socket. Some of those things have pinning requirements, which makes things even harder to get right manually.